crates/rome_js_analyze/src/analyzers/nursery/no_non_null_assertion.rs

```diff
@@ -0,0 +1,178 @@
+        recommended: false,
+    }
+}
+
+fn can_replace_with_optional_chain(node: &TsNonNullAssertionExpression) -> bool {
+    match node.parent::<AnyJsExpression>() {
+        Some(parent) => {
+            matches!(
+                parent.syntax().kind(),
+                JsSyntaxKind::JS_STATIC_MEMBER_EXPRESSION
+                    | JsSyntaxKind::JS_COMPUTED_MEMBER_EXPRESSION
+                    | JsSyntaxKind::JS_CALL_EXPRESSION
+            )
+        }
+        None => false,
+    }
+}
+
+impl Rule for NoNonNullAssertion {
+    type Query = Ast<TsNonNullAssertionExpression>;
+    type State = (TsNonNullAssertionExpression, bool);
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        if let Some(child) = node.syntax().first_child() {
+            let is_child_non_null_assertion =
+                matches!(child.kind(), JsSyntaxKind::TS_NON_NULL_ASSERTION_EXPRESSION);
+            if is_child_non_null_assertion {
+                return None;
+            }
+
+            let most_outer_ts_non_null_assertion_node = node
+                .syntax()
+                .ancestors()
+                .take_while(|n| -> bool {
+                    n.kind() == JsSyntaxKind::TS_NON_NULL_ASSERTION_EXPRESSION
+                })
+                .last()?
+                .cast::<TsNonNullAssertionExpression>()?;
+
+            return Some((
+                most_outer_ts_non_null_assertion_node.clone(),
+                can_replace_with_optional_chain(&most_outer_ts_non_null_assertion_node),
+            ));
+        } else {
+            return Some((node.clone(), can_replace_with_optional_chain(node)));
+        }
+    }
+
+    fn diagnostic(_ctx: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        let (node, _) = state;
+        Some(
+    		RuleDiagnostic::new(
+    			rule_category!(),
+    			node.syntax().text_trimmed_range(),
+    			markup! {
+    				"Forbidden non-null assertion."
+    			},
+    		).note(
+    			markup! {
+    				"Consider using the optional chain operator "<Emphasis>"?."</Emphasis>" instead. This operator includes runtime checks, so it is safer than the compile-only non-null assertion operator.
+                    "
+    			},
+    		)
+    	)
+    }
+
+    fn action(ctx: &RuleContext<Self>, state: &Self::State) -> Option<JsRuleAction> {
+        let (node, can_fix) = state;
+        if !can_fix {
+            return None;
+        }
+
+        let mut mutation = ctx.root().begin();
+
+        node.syntax()
+            .descendants()
+            .map_while(|x| {
+                let child = x.first_child()?;
+                if child.kind() == JsSyntaxKind::TS_NON_NULL_ASSERTION_EXPRESSION {
+                    Some(child.cast::<TsNonNullAssertionExpression>()?)
+                } else {
+                    None
+                }
+            })
```

Comment:
You don't have to use `descendants` here. You can loop over the children directly: 

```suggestion

let mut assertions = std::iter::successors(node.expression().ok(), |expression| match expression {
	AnyJsExpression::TsNonNullAssertionExpression(assertion) => assertion.expression().ok(),
    _ => None
});

// Skip the root node
for assertion in assertions.skip(1) {
	mutation.remove_token(x.excl_token().ok()?);
}
```